File: archive/ruby/muse/util/Sequences.rb
require "muse/algorithms/sorting/QuickSort"
require "muse/util/RandomFactory"

module Sequences
  def self.inspect(arr)
    puts("[")
    arr.each_index do |i|
      print(format("\t#%04X  -->  %d\n", i, arr[i]))
    end
    puts("]")
  end

  def self.isSorted(arr)
    1.upto(arr.length - 1) do |i|
      return false if arr[i] < arr[i - 1]
    end

    true
  end

  def self.parityChecksum(arr)
    checksum = 0

    arr.each do |v|
      checksum ^= v
    end

    checksum
  end

  def self.reverse(arr)
    k = 0

    length = arr.length
    0.upto((arr.length >> 1) - 1) do |i|
      k = length - i - 1
      arr[i], arr[k] = arr[k], arr[i]
    end
  end

  def self.shuffle(arr)
    k = 0
    length = arr.length

    0.upto(arr.length - 1) do |i|
      k = RandomFactory.genIntN(i, length - 1)
      arr[i], arr[k] = arr[k], arr[i]
    end
  end

  def self.sort(arr)
    QuickSort.sort(arr)
  end
end

File: archive/nodejs/muse/util/Sequences.js
import * as QuickSort from '../algorithms/sorting/QuickSort.js';
import * as RandomFactory from './RandomFactory.js';

export const inspect = (arr) => {
  console.log('[');
  for (let i = 0, length = arr.length; i < length; i++) {
    process.stdout.write(
      `\t#${i.toString(16).toUpperCase().padStart(4, '0')}  -->  ${arr[i]}\n`,
    );
  }
  console.log(']');
};

export const isSorted = (arr) => {
  for (let i = 1, length = arr.length; i < length; i++) {
    if (arr[i] < arr[i - 1]) {
      return false;
    }
  }

  return true;
};

export const parityChecksum = (arr) => {
  let checksum = 0;

  for (const v of arr) {
    checksum ^= v;
  }

  return checksum;
};

export const reverse = (arr) => {
  let k;

  for (
    let i = 0, bound = Math.floor(arr.length >>> 1), length = arr.length;
    i < bound;
    i++
  ) {
    k = length - i - 1;
    [arr[i], arr[k]] = [arr[k], arr[i]];
  }
};

export const shuffle = (arr) => {
  let k;

  for (let i = 0, length = arr.length; i < length; i++) {
    k = RandomFactory.genIntN(i, length - 1);
    [arr[i], arr[k]] = [arr[k], arr[i]];
  }
};

export const sort = (arr) => {
  QuickSort.sort(arr);
};
